### PR TITLE
Add eventsource collector and console logger implementation

### LINKED ISSUE #7556: Ship console logger listener with Azure.Core

### PR DESCRIPTION
**Pull request by pakrym**
Adds a way to easily subscribe to all Azure event sources (based on a trait that I would document)
Adds an easy way to get console logging.

The names are subject to change during the API review.

Fixes: https://github.com/Azure/azure-sdk-for-net/issues/7556

### CODE REVIEW

File: sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample4_Logging.cs [hunk 1]
```diff
@@ -24,8 +24,7 @@ public void Logging()
             var client = new ConfigurationClient(connectionString);
 
             // Setup a listener to monitor logged events.
-            var listener = new ConsoleEventListener();
-            listener.EnableEvents(EventLevel.LogAlways);
+            using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger(EventLevel.LogAlways);
```
Comment (schaabs) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample4_Logging.cs]: >using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger(EventLevel.LogAlways); [](start = 12, length = 109)

Should the sample also show overriding the AzureEventSourceListener as well as log to console?

File: sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        private static object FormatValue(object o)
+        {
+            if (o is byte[] bytes)
+            {
+                var stringBuilder = new StringBuilder();
+                foreach (byte b in bytes)
+                {
+                    stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:X2}", b);
+                }
```
Comment (schaabs) [inline sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs]: Why not BitConverter.ToString(data);

**Review by schaabs — APPROVED** [review]

File: sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs (hunk 2 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs]: Format. BitConverter does 00-00-00-00-80 and I don't want the minuses

File: sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample4_Logging.cs (hunk 1 above)
Comment (pakrym) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample4_Logging.cs]: Will do

File: sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                EnableEvents(eventSource, _level);
+            }
+        }
+
+        protected sealed override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            _log(eventData, EventSourceEventFormatting.Format(eventData));
+        }
+
+        public static AzureEventSourceListener CreateConsoleLogger(EventLevel level = EventLevel.Informational)
```
Comment (tg-msft) [inline sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs]: If we're adding conveniences, how about Trace.WriteLine too?

File: sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+namespace Azure.Core.Diagnostics
+{
+    public class AzureEventSourceListener: EventListener
+    {
+        private readonly List<EventSource> _eventSources = new List<EventSource>();
+
+        private readonly Action<EventWrittenEventArgs, string> _log;
+        private readonly EventLevel _level;
+
+        public AzureEventSourceListener(Action<EventWrittenEventArgs, string> log, EventLevel level)
```
Comment (tg-msft) [inline sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs]: Should we create our own type that's `EventWrittenEventArgs + string` with a potentially better name?

File: sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        protected sealed override void OnEventSourceCreated(EventSource eventSource)
+        {
+            base.OnEventSourceCreated(eventSource);
+
+            if (_log == null)
+            {
+                _eventSources.Add(eventSource);
+            }
+
+            if (eventSource.GetTrait("AzureEventSource") == "true")
```
Comment (tg-msft) [inline sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs]: Can we make `"AzureEventSource"` a public constant here?

Comment (tg-msft) [reply sdk/core/Azure.Core/src/Diagnostics/AzureEventSourceListener.cs]: Is it worth doing a `bool.Parse` for safety's sake?

**Review by tg-msft — APPROVED** [review]
Nice!

**Pull request merged**